### PR TITLE
Validate no gaps in UAV store writemask

### LINKED ISSUE #4161: Missing validation for UAV store writemasks that must not contain holes
It seems that DXIL inherited the D3D11 requirements around UAV raw/structured stores, both of which have this text in the D3D11 functional spec:
```
                dst0 can only have a write mask that is one of the
                following: .x, .xy, .xyz, .xyzw.   The writemask determines
                the number of 32bit components to write – without gaps.
```

Attempting to compile HLSL that doesn't follow this restriction will produce multiple stores in the resulting DXIL. Attempting to run DXIL that violates this rule doesn't work correctly. It seems this should be added to DXIL validation for future validator versions.

### PR DESCRIPTION
**Pull request by gracejennings**
This change adds a validation rule for the inherited D3D11 requirement that a UAV store write mask must not contain gaps and therefore be one of the following: .x, .xy, .xyz, .xyzw.

Fixes #4161

### CODE REVIEW

**Review by tex3d — APPROVED** [review]
LGTM!

File: tools/clang/unittests/HLSL/ValidationTest.cpp [hunk 1]
```diff
@@ -1197,6 +1198,16 @@ TEST_F(ValidationTest, UAVStoreMaskMatch) {
...
 
+TEST_F(ValidationTest, UAVStoreMaskGap) {
+  if (m_ver.SkipDxilVersion(1, 6))
+    return;
+  RewriteAssemblyCheckMsg(
+      L"..\\CodeGenHLSL\\uav_store.hlsl", "ps_6_0",
+      "i32 2, i32 2, i32 2, i32 2, i8 15)",
+      "i32 undef, i32 2, i32 undef, i32 2, i8 10)",
+      "UAV write mask must be contiguous and without gaps: .x, .xy, .xyz, or .xyzw.");
+}
```
Comment (pow2clk) [inline tools/clang/unittests/HLSL/ValidationTest.cpp:1209]: I don't think we need to test all 11 invalid combinations, but maybe a couple more such as all set save x (0xE) and only W set (0x8)

File: utils/hct/hctdb.py [hunk 2]
```diff
@@ -2582,6 +2582,7 @@ def build_valrules(self):
         self.add_valrule("Instr.BarrierModeNoMemory", "sync must include some form of memory barrier - _u (UAV) and/or _g (Thread Group Shared Memory).  Only _t (thread group sync) is optional.")
         self.add_valrule("Instr.BarrierModeForNonCS", "sync in a non-Compute/Amplification/Mesh Shader must only sync UAV (sync_uglobal).")
         self.add_valrule("Instr.WriteMaskForTypedUAVStore", "store on typed uav must write to all four components of the UAV.")
+        self.add_valrule("Instr.WriteMaskGapForUAV", "UAV write mask must be contiguous and without gaps: .x, .xy, .xyz, or .xyzw.")
```
Comment (pow2clk) [inline utils/hct/hctdb.py:2585]: I suppose another characteristic that is checked here is that it starts at x. The explicit examples make that clear though.

**Review by pow2clk — COMMENTED** [review]
I'm ambivalent about adding more detail to the validation message. I think with the examples, it's clear what's required, but I would welcome a few more tests just in case

**Review by pow2clk — APPROVED** [review]

**Pull request merged**